Sources/ServiceLifecycle/AsyncCancelOnGracefulShutdownSequence.swift

```diff
@@ -0,0 +1,161 @@
+public struct AsyncCancelOnGracefulShutdownSequence<Base: AsyncSequence & Sendable>: AsyncSequence, Sendable where Base.Element: Sendable {
+    @usableFromInline
+    enum _ElementOrGracefulShutdown: Sendable {
+        case base(AsyncMapNilSequence<Base>.Element)
+        case gracefulShutdown
+    }
+
+    @usableFromInline
+    typealias Merged = AsyncMerge2Sequence<
+        AsyncMapSequence<AsyncMapNilSequence<Base>, _ElementOrGracefulShutdown>,
+        AsyncMapSequence<AsyncMapNilSequence<AsyncGracefulShutdownSequence>, _ElementOrGracefulShutdown>
+    >
+
+    public typealias Element = Base.Element
+
+    @usableFromInline
+    let _merge: Merged
+
+    @inlinable
+    public init(base: Base) {
+        self._merge = .init(
+            base.mapNil().map { .base($0) },
+            AsyncGracefulShutdownSequence().mapNil().map { _ in .gracefulShutdown }
+        )
+    }
+
+    @inlinable
+    public func makeAsyncIterator() -> AsyncIterator {
+        AsyncIterator(iterator: self._merge.makeAsyncIterator())
+    }
+
+    public struct AsyncIterator: AsyncIteratorProtocol {
+        @usableFromInline
+        var _iterator: Merged.AsyncIterator
+
+        @usableFromInline
+        var _isFinished = false
+
+        @inlinable
+        init(iterator: Merged.AsyncIterator) {
+            self._iterator = iterator
+        }
+
+        @inlinable
+        public mutating func next() async rethrows -> Element? {
+            guard !self._isFinished else {
+                return nil
+            }
+
+            let value = try await self._iterator.next()
+
+            switch value {
+            case .base(let element):
+
+                switch element {
+                case .element(let element):
+                    return element
+
+                case .end:
+                    self._isFinished = true
+                    return nil
+                }
+
+            case .gracefulShutdown:
+                return nil
+
+            case .none:
+                return nil
+            }
+        }
+    }
+}
+
+/// This is just a helper extension and sequence to allow us to get the `nil` value as an element of the sequence.
+/// We need this since merge is only finishing when both upstreams are finished but we need to finish when either is done.
+/// In the future, we should move to something in async algorithms if it exists.
+extension AsyncSequence where Self: Sendable, Element: Sendable {
+    @inlinable
+    func mapNil() -> AsyncMapNilSequence<Self> {
+        AsyncMapNilSequence(base: self)
+    }
+}
+
+@usableFromInline
+struct AsyncMapNilSequence<Base: AsyncSequence & Sendable>: AsyncSequence, Sendable where Base.Element: Sendable {
+    @usableFromInline
+    enum ElementOrEnd: Sendable {
+        case element(Base.Element)
+        case end
+    }
+
+    @usableFromInline
+    typealias Element = ElementOrEnd
+
+    @usableFromInline
+    let _base: Base
```

Comment:
many internal variables underscored. why? testing only?

Reply:
This is due to the usage of `@inlinable` which forces us to make these properties `internal` otherwise they would be `private`. We tend to underscore prefix these properties normally to indicate that they are not really `internal`.